File: sdks/go/middleware/handler.go
package middleware

import (
	"context"
	"errors"
	"fmt"
	"html"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	cv "github.com/cvfile/cv-go/cv"
)

// Loader is the pluggable source of `.cv` bytes for a given URL path.
// Returning (nil, nil) means "not found" — return an error only when the
// underlying storage actually failed.
type Loader func(ctx context.Context, path string) ([]byte, error)

// Options tunes the generated handler.
type Options struct {
	// Root serves files from a filesystem directory. Mutually exclusive with Loader.
	Root string
	// Loader, if non-nil, replaces the filesystem lookup entirely.
	Loader Loader
	// CacheControl is the Cache-Control header value. Default: "public, max-age=300".
	CacheControl string
	// DefaultFormat is applied when no Accept and no ?format= query is provided.
	// Empty = "pdf".
	DefaultFormat ServeFormat
}

// Handler returns an http.Handler that serves `.cv` resources with full
// content negotiation.
func Handler(opts Options) (http.Handler, error) {
	if (opts.Root == "") == (opts.Loader == nil) {
		return nil, errors.New("middleware.Handler: provide exactly one of Root or Loader")
	}
	cacheControl := opts.CacheControl
	if cacheControl == "" {
		cacheControl = "public, max-age=300"
	}

	var baseRoot string
	if opts.Root != "" {
		abs, err := filepath.Abs(opts.Root)
		if err != nil {
			return nil, fmt.Errorf("resolve Root %q: %w", opts.Root, err)
		}
		baseRoot = abs
	}

	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet && r.Method != http.MethodHead {
			http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
			return
		}

		formatQuery := r.URL.Query().Get("format")
		if formatQuery == "" && opts.DefaultFormat != "" {
			formatQuery = string(opts.DefaultFormat)
		}

		bytes, err := loadCv(r.Context(), r.URL.Path, baseRoot, opts.Loader)
		if err != nil {
			http.Error(w, "load failed", http.StatusInternalServerError)
			return
		}
		if bytes == nil {
			http.Error(w, "Not found", http.StatusNotFound)
			return
		}
		if !cv.IsCvFile(bytes) {
			http.Error(w, "Not a .cv file", http.StatusUnsupportedMediaType)
			return
		}

		body, contentType, language, format, err := serve(bytes, NegotiateInput{
			Accept:         r.Header.Get("Accept"),
			AcceptLanguage: r.Header.Get("Accept-Language"),
			FormatQuery:    formatQuery,
		})
		if err != nil {
			http.Error(w, "extract failed: "+err.Error(), http.StatusInternalServerError)
			return
		}

		filename := filenameForFormat(r.URL.Path, format)
		header := w.Header()
		header.Set("Content-Type", contentType)
		header.Set("Content-Length", strconv.Itoa(len(body)))
		header.Set("Vary", "Accept, Accept-Language")
		header.Set("Link", BuildLinkHeader(r.URL.Path))
		header.Set("Cache-Control", cacheControl)
		header.Set("Content-Disposition", `inline; filename="`+filename+`"`)
		if language != "" {
			header.Set("Content-Language", language)
		}
		w.WriteHeader(http.StatusOK)
		if r.Method == http.MethodHead {
			return
		}
		_, _ = w.Write(body)
	}), nil
}

func loadCv(ctx context.Context, urlPath, baseRoot string, loader Loader) ([]byte, error) {
	if loader != nil {
		return loader(ctx, urlPath)
	}
	if baseRoot == "" {
		return nil, nil
	}
	rel := filepath.Clean(strings.TrimLeft(urlPath, "/\\"))
	full := filepath.Join(baseRoot, rel)
	abs, err := filepath.Abs(full)
	if err != nil {
		return nil, nil
	}
	if abs != baseRoot && !strings.HasPrefix(abs, baseRoot+string(os.PathSeparator)) {
		return nil, nil // path traversal — silently 404
	}
	info, err := os.Stat(abs)
	if err != nil || info.IsDir() {
		return nil, nil
	}
	return os.ReadFile(abs)
}

func serve(cvBytes []byte, in NegotiateInput) (body []byte, contentType string, language string, format ServeFormat, err error) {
	decision := Negotiate(in)
	if decision.Format == FormatPDF {
		return cvBytes, PDFPrimaryMIME, decision.Language, FormatPDF, nil
	}

	file, err := cv.Extract(cvBytes)
	if err != nil {
		return nil, "", "", "", err
	}
	preferLang := decision.Language
	if preferLang == "" {
		preferLang = file.Metadata.PrimaryLanguage
	}

	if decision.Format == FormatMarkdown {
		md := pickPayload(file, "text/markdown", preferLang)
		if md != nil {
			lang := md.Language
			if lang == "" {
				lang = preferLang
			}
			return md.Bytes, "text/markdown; charset=utf-8; cv-language=" + lang, md.Language, FormatMarkdown, nil
		}
		return cvBytes, PDFPrimaryMIME, decision.Language, FormatPDF, nil
	}

	if decision.Format == FormatHTML {
		htmlPayload := pickPayload(file, "text/html", preferLang)
		if htmlPayload != nil {
			lang := htmlPayload.Language
			if lang == "" {
				lang = preferLang
			}
			return htmlPayload.Bytes, "text/html; charset=utf-8; cv-language=" + lang, htmlPayload.Language, FormatHTML, nil
		}
		md := pickPayload(file, "text/markdown", preferLang)
		if md != nil {
			rendered := renderMarkdownAsHTML(string(md.Bytes), file)
			return []byte(rendered), "text/html; charset=utf-8", md.Language, FormatHTML, nil
		}
		return cvBytes, PDFPrimaryMIME, decision.Language, FormatPDF, nil
	}

	return cvBytes, PDFPrimaryMIME, decision.Language, FormatPDF, nil
}

func pickPayload(file *cv.File, mime, preferLang string) *cv.ExtractedPayload {
	var first *cv.ExtractedPayload
	for i := range file.Payloads {
		p := &file.Payloads[i]
		if p.MimeType != mime {
			continue
		}
		if p.Language == preferLang {
			return p
		}
		if first == nil {
			first = p
		}
	}
	return first
}

func filenameForFormat(urlPath string, format ServeFormat) string {
	base := urlPath
	if i := strings.LastIndex(base, "/"); i >= 0 {
		base = base[i+1:]
	}
	if base == "" {
		base = "document"
	}
	stem := base
	for _, suffix := range []string{".cv", ".pdf", ".md", ".html"} {
		if strings.HasSuffix(strings.ToLower(stem), suffix) {
			stem = stem[:len(stem)-len(suffix)]
			break
		}
	}
	if stem == "" {
		stem = "document"
	}
	switch format {
	case FormatMarkdown:
		return stem + ".md"
	case FormatHTML:
		return stem + ".html"
	default:
		return stem + ".cv"
	}
}

func renderMarkdownAsHTML(md string, file *cv.File) string {
	safe := html.EscapeString(md)
	lang := file.Metadata.PrimaryLanguage
	title := html.EscapeString(file.Metadata.PrimaryPayload)
	return "<!doctype html>\n" +
		`<html lang="` + lang + `">` + "\n" +
		"<head>\n" +
		`<meta charset="utf-8">` + "\n" +
		"<title>" + title + "</title>\n" +
		"</head>\n" +
		"<body>\n" +
		"<pre>" + safe + "</pre>\n" +
		"</body>\n" +
		"</html>"
}

File: sdks/python/src/cvfile/server/_handler.py
"""Framework-agnostic core: bytes in, bytes + headers out.

ASGI/WSGI adapters wrap this. Mirrors @cvfile/server's serve.ts +
handler.ts so a request flowing through Python returns the same body
and headers Node would return for the same Accept.
"""

from __future__ import annotations

import html as _html
import sys
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, NoReturn

if sys.version_info >= (3, 11):
    from typing import assert_never
else:  # assert_never landed in typing only in Python 3.11; we support 3.10+.

    def assert_never(value: object, /) -> NoReturn:
        raise AssertionError(f"Unhandled value: {value!r}")

from cvfile.extract import extract
from cvfile.server._conneg import (
    PDF_PRIMARY_MIME,
    NegotiationResult,
    ServeFormat,
    build_link_header,
    negotiate,
)

if TYPE_CHECKING:
    from cvfile._types import CvFile, ExtractedPayload

CACHE_CONTROL_DEFAULT = "public, max-age=300"


@dataclass(frozen=True, slots=True)
class ServeOptions:
    cache_control: str = CACHE_CONTROL_DEFAULT
    default_format: str | None = None


@dataclass(frozen=True, slots=True)
class ServeBody:
    format: ServeFormat
    content_type: str
    body: bytes
    language: str | None = None


@dataclass(frozen=True, slots=True)
class ServeResponse:
    status: int
    body: bytes
    headers: dict[str, str] = field(default_factory=dict)


def serve_cv_bytes(
    cv_bytes: bytes,
    *,
    accept: str | None = None,
    accept_language: str | None = None,
    format_query: str | None = None,
    self_url: str = "/",
    logical_name: str = "document",
    options: ServeOptions | None = None,
) -> ServeResponse:
    """Negotiate, extract, and return a fully-formed HTTP response for a `.cv`."""
    opts = options or ServeOptions()
    decision = negotiate(
        accept=accept,
        accept_language=accept_language,
        format_query=format_query or opts.default_format,
    )

    body = _serve(cv_bytes, decision)
    headers: dict[str, str] = {
        "Content-Type": body.content_type,
        "Content-Length": str(len(body.body)),
        "Vary": "Accept, Accept-Language",
        "Link": build_link_header(self_url, cv_mime=PDF_PRIMARY_MIME),
        "Cache-Control": opts.cache_control,
        "Content-Disposition": f'inline; filename="{_filename_for(logical_name, body.format)}"',
    }
    if body.language:
        headers["Content-Language"] = body.language

    return ServeResponse(status=200, body=body.body, headers=headers)


def _serve(cv_bytes: bytes, decision: NegotiationResult) -> ServeBody:
    if decision.format == "pdf":
        return ServeBody(
            format="pdf",
            content_type=PDF_PRIMARY_MIME,
            body=cv_bytes,
            language=decision.language,
        )

    file = extract(cv_bytes)
    prefer_lang = decision.language or file.metadata.primary_language

    if decision.format == "markdown":
        md = _pick(file, "text/markdown", prefer_lang)
        if md is not None:
            lang = md.language or prefer_lang
            return ServeBody(
                format="markdown",
                content_type=f"text/markdown; charset=utf-8; cv-language={lang}",
                body=md.bytes_,
                language=md.language,
            )
        return _fallback_pdf(cv_bytes)

    if decision.format == "html":
        html = _pick(file, "text/html", prefer_lang)
        if html is not None:
            lang = html.language or prefer_lang
            return ServeBody(
                format="html",
                content_type=f"text/html; charset=utf-8; cv-language={lang}",
                body=html.bytes_,
                language=html.language,
            )
        md = _pick(file, "text/markdown", prefer_lang)
        if md is not None:
            rendered = _render_markdown_as_html(md.bytes_.decode("utf-8", errors="replace"), file)
            return ServeBody(
                format="html",
                content_type="text/html; charset=utf-8",
                body=rendered.encode("utf-8"),
                language=md.language,
            )
        return _fallback_pdf(cv_bytes)

    assert_never(decision.format)


def _pick(file: CvFile, mime: str, prefer_lang: str) -> ExtractedPayload | None:
    matches = [p for p in file.payloads if p.mime_type == mime]
    if not matches:
        return None
    for p in matches:
        if p.language == prefer_lang:
            return p
    return matches[0]


def _fallback_pdf(cv_bytes: bytes) -> ServeBody:
    return ServeBody(format="pdf", content_type=PDF_PRIMARY_MIME, body=cv_bytes)


def _filename_for(logical: str, fmt: ServeFormat) -> str:
    base = logical.rsplit("/", 1)[-1] or "document"
    stem = base
    for suffix in (".cv", ".pdf", ".md", ".html"):
        if stem.lower().endswith(suffix):
            stem = stem[: -len(suffix)]
            break
    if not stem:
        stem = "document"
    if fmt == "markdown":
        return f"{stem}.md"
    if fmt == "html":
        return f"{stem}.html"
    return f"{stem}.cv"


def _render_markdown_as_html(md: str, file: CvFile) -> str:
    safe = _html.escape(md)
    lang = file.metadata.primary_language
    title = _html.escape(file.metadata.primary_payload)
    return (
        "<!doctype html>\n"
        f'<html lang="{lang}">\n'
        "<head>\n"
        '<meta charset="utf-8">\n'
        f"<title>{title}</title>\n"
        "</head>\n"
        "<body>\n"
        f"<pre>{safe}</pre>\n"
        "</body>\n"
        "</html>"
    )
